### PR TITLE
Added TF Model support

### PR DESCRIPTION
**Pull request by juntai-zheng**
Added support for TF Model saving and loading. I based the code and tests on `mlflow/sklearn.py` and `mlflow/tests/mlflow/sklearn/test_sklearn_model_export.py`. 

The main difference between `tensorflow` and `sklearn` is that `tensorflow` needs to save both a `model_fn` file and `model_dir` file. From that we can reconstruct a `tf.estimator.Estimator` based on those and predict on a `pandas.DataFrame input`. Included in `tensorflow.py` is `TFWrapper`, which handles converting the input from a `DataFrame` to an input function for an estimator. It also handles converting the output of the estimator's prediction to be a `DataFrame` as well.

### CODE REVIEW

File: mlflow/tensorflow.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import tensorflow as tf
+
+from mlflow.utils.file_utils import TempDir
+from mlflow import pyfunc
+from mlflow.models import Model
+import mlflow.tracking
+
+
+class TFWrapper(object):
+    """Wrapper class that creates a predict function such that
```
Comment (smurching) [inline mlflow/tensorflow.py]: Style nit: Start multiline docstrings on a new line, e.g.
```
"""
Wrapper class that...
"""

File: mlflow/tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+class TFWrapper(object):
+    """Wrapper class that creates a predict function such that 
+    predict(data: pandas.DataFrame) -> pandas.DataFrame"""
+    def __init__(self, model_fn, model_dir):
+        self.estimator = tf.estimator.Estimator(model_fn, model_dir=model_dir)
+
+    def predict(self, df):
+        input_fn = tf.estimator.inputs.pandas_input_fn(df, shuffle=False)
+        pred = self.estimator.predict(input_fn)
+        results = []
```
Comment (smurching) [inline mlflow/tensorflow.py]: You can do this with a list comprehension, e.g.:
```
results = [p['predictions'] for p in self.estimator.predict(input_fn)]
```

File: mlflow/tensorflow.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        return pandas.DataFrame(results)
+
+
+def save_model(tf_model, path, conda_env=None, mlflow_model=Model()):
+    """Save a Tensorflow model to a directory in the local file system."""
+    if os.path.exists(path):
+        raise Exception("Path '{}' already exists".format(path))
+    os.makedirs(path)
+    model_fn_file = os.path.join(path, "model_fn.pkl")
+    model_dir_file = os.path.join(path, "model_dir.pkl")
```
Comment (smurching) [inline mlflow/tensorflow.py]: `model_dir` is a string (path to the estimator's model directory) right? It'd be good to directly copy the model directory into the directory at `path`. That way you could share your exported model directory with me & I could load the model back (I'd otherwise end up trying to read from a directory on your local filesystem).

File: mlflow/tensorflow.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    model_dir = None
+    with open(os.path.join(path, params["dir"]), "rb") as f:
+        model_dir = pickle.load(f)
+    return TFWrapper(model_fn, model_dir)
+
+
+def load_pyfunc(path):
+    model_fn = None
+    model_dir = None
+    for filename in os.listdir(path):
```
Comment (smurching) [inline mlflow/tensorflow.py]: Would be nice to explicitly look for `model_fn.pkl` & raise an Exception if it doesn't exist. Possible error message: `"Could not find pickled TensorFlow model function file %s in directory %s" % ("model_fn.pkl", path)`.

File: tests/mlflow/tensorflow/test_tensorflow_model_export.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            x = tensorflow.load_model(path)
+            xpred = x.predict(pandas.DataFrame(data=self._X, columns=self._feature_names))
+            saved = []
+            for s in self._dnn_predict:
+                saved.append(s['predictions'])
+            loaded = []
+            for index, rows in xpred.iterrows():
+                loaded.append(rows)
+            np.testing.assert_array_equal(saved, loaded)
+            # sklearn should also be stored as a valid pyfunc model
```
Comment (smurching) [inline tests/mlflow/tensorflow/test_tensorflow_model_export.py]: typo: sklearn -> TensorFlow estimator

File: tests/mlflow/tensorflow/test_tensorflow_model_export.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        with TempDir(chdr=True, remove_on_exit=True) as tmp:
+            path = tmp.path("dnn")
+            tensorflow.save_model(tf_model=self._dnn, path=path)
+            x = tensorflow.load_model(path)
+            xpred = x.predict(pandas.DataFrame(data=self._X, columns=self._feature_names))
+            saved = []
+            for s in self._dnn_predict:
+                saved.append(s['predictions'])
+            loaded = []
+            for index, rows in xpred.iterrows():
```
Comment (smurching) [inline tests/mlflow/tensorflow/test_tensorflow_model_export.py]: Could we actually convert the original model's predictions into a DataFrame & compare the loaded model's predictions (`xpred`) to the expected DataFrame? That way we more directly check that `x.predict` gives us what we want (a DataFrame)

File: mlflow/tensorflow.py (hunk 4 above)
Comment (smurching) [reply mlflow/tensorflow.py]: Also an observation for other reviewers: looks like TensorFlow estimators create a random temp directory to use as a model directory if it's unspecified, so we don't need to worry about the directory being `None`

**Review by smurching — COMMENTED** [review]
Thanks @juntai-zheng! Leaving a few comments

File: mlflow/tensorflow.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import pandas
+import tensorflow as tf
+
+from mlflow.utils.file_utils import TempDir
+from mlflow import pyfunc
+from mlflow.models import Model
+import mlflow.tracking
+
+
+class TFWrapper(object):
```
Comment (mateiz) [inline mlflow/tensorflow.py]: Is this supposed to be a public API? If not, prefix it with `_`

File: mlflow/tensorflow.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    def predict(self, df):
+        input_fn = tf.estimator.inputs.pandas_input_fn(df, shuffle=False)
+        pred = self.estimator.predict(input_fn)
+        results = []
+        for p in pred:
+            results.append(p['predictions'])
+        return pandas.DataFrame(results)
+
+
+def save_model(tf_model, path, conda_env=None, mlflow_model=Model()):
```
Comment (mateiz) [inline mlflow/tensorflow.py]: Why do we need `conda_env`? So people can specify the TF version?

File: tests/mlflow/tensorflow/test_tensorflow_model_export.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        self._tmp = tempfile.mkdtemp()
+        iris = datasets.load_iris()
+        self._X = iris.data[:, :2]  # we only take the first two features.
+        self._y = iris.target
+        self._trainingFeatures = {}
+        self._feature_names = iris.feature_names[:2]
+        for i in range(0, 2):
+            iris.feature_names[i] = iris.feature_names[i].replace(" ", "")
+            iris.feature_names[i] = iris.feature_names[i].replace("(", "")
+            iris.feature_names[i] = iris.feature_names[i].replace(")", "")
```
Comment (mateiz) [inline tests/mlflow/tensorflow/test_tensorflow_model_export.py]: Why do we need these? add a comment

File: mlflow/cli.py [hunk 10]
```diff
@@ -132,6 +132,7 @@ def ui(file_store_path):
     server.app.run("0.0.0.0")
 
 
+cli.add_command(mlflow.tensorflow.commands)
```
Comment (smurching) [inline mlflow/cli.py]: We should remove this since we aren't adding any TensorFlow-specific CLI commands (?)

File: tests/mlflow/tensorflow/test_tensorflow_model_export.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import numpy as np
+import tensorflow as tf
+import sklearn.datasets as datasets
+
+from mlflow import tensorflow, pyfunc
+from mlflow import tracking
+from mlflow.utils.file_utils import TempDir
+
+
+class TestModelExport(unittest.TestCase):
```
Comment (smurching) [inline tests/mlflow/tensorflow/test_tensorflow_model_export.py]: Based on conversation with @andrewmchen & @tomasatdatabricks we should refactor this to to pytest format, though IMO we can wait till after we release TF functionality.

Some notes on how this could be done:
* Use the [tmpdir fixture](https://docs.pytest.org/en/latest/tmpdir.html) to generate a temporary test dir.
* See https://github.com/databricks/mlflow/blob/master/tests/tracking/test_tracking.py for an example of pytest-style tests (except I didn't / should have used the tmpdir fixture when writing them :P). 
* Instance variables used in the test method (`self._dnn`, `self._dnn_predict`, `self._X`, `self._feature_names`) can be passed as fixtures. We could have a utility method that returns a tuple of these values & use [pytest.mark.parametrize](https://docs.pytest.org/en/latest/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions) to feed values into the test function. Speaking of which, we should be able to implement the test with just `estimator`, `X`, and `y` fixture values (fit the estimator on X and y, compare predictions on y before & after saving/loading the model).

**Review by smurching — COMMENTED** [review]
Had a few more comments, https://github.com/databricks/mlflow/pull/59 will hopefully unblock this PR & then after this round of comments we can merge :)

**Comment by juntai-zheng** [comment]
To avoid generating artifacts in the current `./mlruns `directory, a run was required to start (`tracking.start_run()`) due to `tracking.set_tracking_uri()` not taking effect until a new run gets started. It seems like other tests in the test suite are creating runs, however.

File: tox-requirements.txt [hunk 12]
```diff
@@ -1,11 +1,13 @@
 # Test reqs
+dill
```
Comment (smurching) [inline tox-requirements.txt]: Nit: do we still need `dill`?

**Review by smurching — APPROVED** [review]
LGTM merging, thanks @juntai-zheng! We'll make follow-ups to add documentation & a test verifying PyFunc inference works with non-numeric data types.
